Mac stack build to prevent exception throwing causing aborts

srherbener opened the pull request:
It turns out that the MPI builds are critical to getting rid of the aborts (see discussion in JCSDA/oops#649). The issue comes from the JEDI having C/C++/Fortran mixed source code and the Mac using Apple Clang for C/C++ compilation and GNU Fortran for Fortran compilation. Apple Clang and GNU have each their own implementations of the C++ std::library and getting the MPI wrappers (mpicc, mpicxx, mpifort) built to use the proper link step options is key.

This PR modifies the Mac builds so that the proper settings for for the MPI wrappers are used.

I also updated the JCSDA forks to use the release-stable branches of the ECMWF forked repositories, as well as the anticipated pnetcdf/netcdf/hdf5 version updates.

ytremolet commented:
I am getting the following error from the setup_modules.sh run:
```
checking for linker for Fortran main program... configure: error: Could not determine a way to link a Fortran test program!
+ [[ N =~ [yYtT] ]]
+ set +x
setup_modules.sh custom: success!
```
I can figure out what's happening but it would be nice if the final line doesn't say success because I didn't notice the error at first.

srherbener commented:
@ytremolet, would it be okay to address the issue of writing "success" when there is an error in a subsequent PR? If so, we could file a ZenHub issue so we keep track of it.

The error message you posted is suggesting that you ran ``setup_moduels.sh custom``. For the Mac you should use the argument "mac" instead of "custom": ``setup_modules.sh mac``. Same for the subsequent build_stack command: ``build_stack.sh mac``.

Note the corresponding configuration is in ``jedi-stack/buildscripts/config/confg_mac.sh`` if you want to change anything.

ytremolet commented:
I managed to build mpich with the `setup_modules.sh` script. However, there are a few issues:
- In the instructions, it says not to run `setup_environment.sh` for the mac, but it's missing something about running `module use /opt/modules/...`
- There are errors in the way it is defining MPI_ROOT and a few other variables which it points to a location that does not exist. It seems to be missing part of the path, or it did not set a link the way it was supposed to, I am not sure which.

srherbener commented:
@ytremolet, thanks for the feedback. I'll take a look at the issues you have raised, including the faulty "success" message. 

Could you send me more details about the MPI_ROOT problem. Are you saying that when you do a ``module load ...`` that MPI_ROOT is set incorrectly? I've been using ``module load jedi/clang-mpich`` which seems to work on my Macs.

mmiesch commented:
I am building now and I did not run into the MPI_ROOT problem.  setup_modules ran smoothly and now I'm building the stack - currently on hdf5.    Looks good so far.  I'm running with clang 11.0.3 and gfortran 9.3.0 on Catalina (10.15.4)

ytremolet commented:
Here is what I see after `module load jedi-mpich`:
```
echo $MPI_ROOT
/opt/modules/clang-11.0.3/mpich/3.3.2

ls: /opt/modules/clang-11.0.3/mpich/3.3.2: No such file or directory
```

srherbener commented:
When I try the following in a fresh window, it works on my MacBook:
```
module purge
module load jedi-clang
module load jedi-mpich
module list
  Currently Loaded Modules:
    1) jedi-clang/11.0.0   2) szip/2.1.1   3) mpich/3.3.1   4) jedi-mpich/3.3.1
echo $MPI_ROOT
  /opt/modules/clang-11.0.0/mpich/3.3.1
ll -d $MPI_ROOT
  drwxr-xr-x  6 root  staff  192 May 21 16:16 /opt/modules/clang-11.0.0/mpich/3.3.1
```

When I built mpich, using jedi-stack, I ran ``setup_modules.sh mac`` and I have the following settings in my buildscripts/config/config_mac.sh file:

```
# Compiler/MPI combination
export JEDI_COMPILER="clang/11.0.0"
export JEDI_MPI="mpich/3.3.1"
```

These are the default settings. It looks like @ytremolet you are using slightly newer versions which should be okay. I believe that all you need to do is make sure the JEDI_COMPILER and JEDI_MPI are set to the proper values in the config_mac.sh script.

@ytremolet what does ``ls /opt/modules/clang-11.0.3`` return on your system. If that shows an ``mpich`` directory, then what does ``ls /opt/modules/clang-11.03/mpich`` return.

Also there might be some error messages in the jedi-stack/buildscripts/log/mpich.log file. Maybe the build got an error but the setup_modules.sh script erroneously reported "success".

mmiesch commented on buildscripts/build_stack.sh:
```diff
@@ -80,7 +80,7 @@ build_lib LAPACK lapack 3.7.0
 build_lib EIGEN3 eigen 3.3.5
 build_lib BUFRLIB bufrlib master
-build_lib ECBUILD ecbuild jcsda 3.3.0.jcsda3
+build_lib ECBUILD ecbuild jcsda release-stable
```
I would prefer to leave the tagged versions in jedi-stack for use with containers and environment modules.  A module called `release-stable` doesn't tell you anything about how up to date it is.

srherbener replied:
I'm good with that. I'll move the modules I changed to release-stable back to specific tags.

mmiesch replied:
Thanks - please use jcsda 1.10.1.jcsda2 for eckit.  I just defined that that - it has the hotfix I mentioned this morning for intel 19+ compilers.

mmiesch commented on buildscripts/build_stack.sh:
```diff
@@ -89,7 +89,7 @@ build_lib HDF5 hdf5 1.10.6
 build_lib NCCMP nccmp 1.8.6.5
-build_lib ECKIT eckit jcsda 1.10.1.jcsda1
+build_lib ECKIT eckit jcsda 1.10.1.jcsda2
 build_lib FCKIT fckit jcsda develop
```
Thanks @srherbener for changing back.  Just one minor comment - I think this was a bug before - fckit should probably be release-stable.  But, since we don't typically build it with the stack it't not a big concern.

srherbener replied:
I just changed fckit to the release-stable branch. I also verified that STACK_BUILD_FCKIT is set to 'N' in all of the config scripts, so this change shouldn't thrash anything.

srherbener commented:
I think I've got all the reviewer comments addressed.

- Changed the build_stack.sh versions back to specific tags

- Added the missing instruction (module use ...) to the README.md documentation

- Put in code in setup_modules.sh to properly catch and exit errors from szip and mpi builds

I have not been able to reproduce the MPI_ROOT issue that @ytremolet reported. Has this been resolved, or is it still an issue?
